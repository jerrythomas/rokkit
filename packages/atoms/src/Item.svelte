<script>
	import Icon from './Icon.svelte'
	import { FieldMapper } from '@rokkit/core'

	/**
	 * @type {Object} Props
	 * @property {string|Object} [value]
	 * @property {FieldMapper}   [mapping]
	 */
	let { value, mapping = new FieldMapper() } = $props()
</script>

{#if mapping.hasIcon(value)}
	<Icon name={mapping.getIcon(value)} label={mapping.getLabel(value)} />
{:else if mapping.hasImage(value)}
	<img src={mapping.getImage(value)} alt={mapping.getLabel(value)} />
{/if}
<p>{mapping.getText(value)}</p>
