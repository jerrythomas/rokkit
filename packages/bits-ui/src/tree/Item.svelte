<script>
	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 */

	/** @type {Props & { [key: string]: any }} */
	let { class: className = '', children, ...rest } = $props()
</script>

<tree-item {...rest} class={className} data-tree-item>
	{@render children?.()}
</tree-item>
