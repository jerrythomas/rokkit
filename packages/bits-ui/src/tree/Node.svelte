<script>
	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any[]} [items]
	 */

	/** @type {Props & { [key: string]: any }} */
	let { class: className = '', children, nodeIcon, items = [], ...rest } = $props()
	let lines = []
	let hasChildren = $derived(Array.isArray(items) && items.length > 0)
</script>

<tree-node {...rest} class={className} data-tree-node>
	{#each lines as { type }, index (index)}
		<Line {type} />
	{/each}
	{#if hasChildren}
		{@render nodeIcon(expanded)}
	{/if}
	{@render children?.()}
</tree-node>
