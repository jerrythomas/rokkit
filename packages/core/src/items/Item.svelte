<script>
	import { defaultFields } from '../constants'
	import Icon from '../Icon.svelte'

	export let value
	export let fields = defaultFields

	$: isObject = typeof value == 'object'
	$: text = isObject ? value[fields.text] : value
</script>

{#if isObject}
	{#if value[fields.icon]}
		{@const iconName =
			typeof value[fields.icon] == 'object'
				? value[fields.icon][value[fields.state]]
				: value[fields.icon]}
		<Icon name={iconName} />
	{:else if value[fields.image]}
		<img
			class="h-4 w-4 object-cover"
			alt={value[fields.text]}
			src={value[fields.image]}
		/>
	{/if}
{/if}
{#if text}
	<p class="w-full flex">{text}</p>
{/if}
