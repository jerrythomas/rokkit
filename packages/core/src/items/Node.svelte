<script>
	import Connector from './Connector.svelte'
	import { defaultFields, defaultStateIcons } from '../constants'

	export let value
	export let fields = defaultFields
	export let types = []
	export let stateIcons = defaultStateIcons.node

	export let selected = false
	export let using = {}
	export let rtl = false
	export let path = []

	$: stateIcons = { ...defaultStateIcons.node, ...(stateIcons ?? {}) }
	$: hasChildren = fields.children in value
	$: state =
		hasChildren && value[fields.isOpen]
			? { icon: stateIcons.opened, label: 'collapse' }
			: { icon: stateIcons.closed, label: 'expand' }
	$: component = value[fields.component]
		? using[value[fields.component]] || using.default
		: using.default
</script>

<!-- svelte-ignore a11y-click-events-have-key-events -->
<node
	id={'id-' + path.join('-')}
	class="flex flex-row min-h-5 items-center cursor-pointer select-none"
	class:is-selected={selected}
	class:flex-row-reverse={rtl}
	aria-selected={selected}
	role="option"
	data-path={path.join(',')}
>
	{#each types as type}
		{#if type === 'icon'}
			<span class="flex flex-col w-4 h-full items-center justify-center">
				<icon class={state.icon} aria-label={state.label} tabindex="-1" />
			</span>
		{:else}
			<Connector {type} />
		{/if}
	{/each}
	<item>
		<svelte:component this={component} bind:value {fields} />
	</item>
</node>
