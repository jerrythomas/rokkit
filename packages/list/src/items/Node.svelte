<script>
	import Connector from './Connector.svelte'
	import { Icon } from '@sparsh-ui/icons'

	export let content
	export let fields = {}
	export let types = []
	export let hasChildren = false
	export let linesVisible

	function toggle() {
		content.isOpen = !content.isOpen
	}
</script>

<node
	class="flex flex-row h-8 gap-2 leading-loose items-center cursor-pointer select-none"
	on:click={toggle}
>
	{#each types.slice(1) as type}
		<Connector type={linesVisible ? type : 'empty'} />
	{/each}

	{#if hasChildren}
		<span class="flex flex-col w-4 h-full items-center justify-center">
			{#if content.isOpen}
				<Icon size="100%" name="node-open" title="Collapse" />
			{:else}
				<Icon size="100%" name="node-closed" title="Expand" />
			{/if}
		</span>
	{/if}
	{#if content[fields.image]}
		<img
			class="h-8 w-8 rounded-full"
			alt={content[fields.text]}
			src={content[fields.image]}
		/>
	{/if}
	{#if content[fields.icon]}
		<Icon name={content[fields.icon]} title={content[fields.text]} />
	{/if}

	<p class="flex flex-grow">{content[fields.text]}</p>
</node>
