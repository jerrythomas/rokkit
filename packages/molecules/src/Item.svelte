<script>
	import { defaultFields, getIcon, getText } from '@rokkit/core'
	import { Icon } from '@rokkit/atoms'

	export let value
	export let fields = defaultFields

	$: isObject = typeof (value ?? '-') == 'object'
	$: text = getText(value, fields)
</script>

{#if isObject}
	{#if value[fields.icon]}
		{@const iconName = getIcon(value, fields)}
		<Icon name={iconName} />
	{:else if value[fields.image]}
		<img class="min-h-4 min-w-4 object-cover" alt={value[fields.text]} src={value[fields.image]} />
	{/if}
{/if}
{#if text}
	<p>{text}</p>
{/if}
