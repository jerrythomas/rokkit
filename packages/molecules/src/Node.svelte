<script>
	import { defaultStateIcons } from '@rokkit/core'
	import { defaultMapping } from './constants'
	import { Icon, Connector } from '@rokkit/atoms'

	/**
	 * @typedef {Object} Props
	 * @property {any} value
	 * @property {import('@rokkit/core').FieldMapper} [mapping]
	 * @property {any} [types]
	 * @property {import('./types').NodeStateIcons} [stateIcons]
	 * @property {boolean} [selected]
	 * @property {boolean} [expanded]
	 * @property {number[]} [path]
	 * @property {import('svelte').Snippet} [children]
	 */

	/** @type {Props} */
	let {
		class: classes = '',
		value = $bindable(),
		mapping = defaultMapping,
		types = [],
		stateIcons = defaultStateIcons.node,
		selected = false,
		expanded = false,
		path = [],
		children
	} = $props()

	let icons = $derived({ ...defaultStateIcons.node, ...stateIcons })
	let stateName = $derived(expanded ? 'opened' : 'closed')
	let state = $derived(
		expanded ? { icon: icons.opened, label: 'collapse' } : { icon: icons.closed, label: 'expand' }
	)

	const Template = $derived(mapping.getComponent(value))
	const id = $derived('id-' + path.join('-'))
</script>

<rk-node
	{id}
	class={classes}
	aria-selected={selected}
	aria-expanded={state.label === 'collapse'}
	role="treeitem"
	data-path={path.join(',')}
>
	<div class="flex flex-row items-center">
		{#each types as type}
			{#if type === 'icon'}
				<Icon name={state.icon} label={state.label} state={stateName} class="w-4" size="small" />
			{:else}
				<Connector {type} />
			{/if}
		{/each}
		<rk-item>
			<Template bind:value {mapping} />
		</rk-item>
	</div>
	{@render children?.()}
</rk-node>
