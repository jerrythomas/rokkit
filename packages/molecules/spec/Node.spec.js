import { describe, expect, it, beforeEach } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import { tick } from 'svelte'
import Node from '../src/Node.svelte'
import Item from '../src/Item.svelte'

describe('Node.svelte', () => {
	const using = { default: Item }
	beforeEach(() => cleanup())
	it('should render node', () => {
		const { container } = render(Node, { value: { text: 'Alpha' }, using })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render node with path', () => {
		const { container } = render(Node, {
			value: { text: 'Example', path: '/alpha' },
			using,
			path: [0]
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render node with collapsed icon', () => {
		const { container } = render(Node, {
			value: { text: 'Collapsed Node', children: ['A', 'B'], types: ['icon'] },
			using
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})
	it('should render node with expanded icon', () => {
		const { container } = render(Node, {
			value: {
				text: 'Expanded Node',
				children: ['A', 'B'],
				types: ['icon'],
				_open: true
			},
			using
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})
	it('should render node sibling node', () => {
		const { container } = render(Node, {
			value: { text: 'Sibling Node', types: ['sibling'] },
			using
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render node when selected', () => {
		const { container } = render(Node, {
			value: { text: 'A Selected Node' },
			selected: true,
			using
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render node in rtl', () => {
		const { container } = render(Node, {
			value: { text: 'An RTL Node' },
			rtl: true,
			using
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render node when path changes', async () => {
		const { container, component } = render(Node, {
			value: { text: 'A Node' },
			path: [0],
			using
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
		component.$set({ path: [1] })
		await tick()
		expect(container).toMatchSnapshot()
	})
})
