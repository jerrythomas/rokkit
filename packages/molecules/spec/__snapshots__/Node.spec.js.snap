// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Node.svelte > should render node 1`] = `
<body>
  <div>
    <node
      aria-selected="false"
      class="min-h-5 flex flex-row cursor-pointer select-none items-center"
      data-path=""
      id="id-"
      role="option"
    >
       
      <item>
        
         
        <p>
          Alpha
        </p>
        
      </item>
    </node>
  </div>
</body>
`;

exports[`Node.svelte > should render node in rtl 1`] = `
<body>
  <div>
    <node
      aria-selected="false"
      class="min-h-5 flex flex-row cursor-pointer select-none items-center flex-row-reverse"
      data-path=""
      id="id-"
      role="option"
    >
       
      <item>
        
         
        <p>
          An RTL Node
        </p>
        
      </item>
    </node>
  </div>
</body>
`;

exports[`Node.svelte > should render node sibling node 1`] = `
<body>
  <div>
    <node
      aria-selected="false"
      class="min-h-5 flex flex-row cursor-pointer select-none items-center"
      data-path=""
      id="id-"
      role="option"
    >
       
      <item>
        
         
        <p>
          Sibling Node
        </p>
        
      </item>
    </node>
  </div>
</body>
`;

exports[`Node.svelte > should render node when path changes 1`] = `
<body>
  <div>
    <node
      aria-selected="false"
      class="min-h-5 flex flex-row cursor-pointer select-none items-center"
      data-path="0"
      id="id-0"
      role="option"
    >
       
      <item>
        
         
        <p>
          A Node
        </p>
        
      </item>
    </node>
  </div>
</body>
`;

exports[`Node.svelte > should render node when path changes 2`] = `
<body>
  <div>
    <node
      aria-selected="false"
      class="min-h-5 flex flex-row cursor-pointer select-none items-center"
      data-path="1"
      id="id-1"
      role="option"
    >
       
      <item>
        
         
        <p>
          A Node
        </p>
        
      </item>
    </node>
  </div>
</body>
`;

exports[`Node.svelte > should render node when selected 1`] = `
<body>
  <div>
    <node
      aria-selected="true"
      class="min-h-5 flex flex-row cursor-pointer select-none items-center is-selected"
      data-path=""
      id="id-"
      role="option"
    >
       
      <item>
        
         
        <p>
          A Selected Node
        </p>
        
      </item>
    </node>
  </div>
</body>
`;

exports[`Node.svelte > should render node with collapsed icon 1`] = `
<body>
  <div>
    <node
      aria-selected="false"
      class="min-h-5 flex flex-row cursor-pointer select-none items-center"
      data-path=""
      id="id-"
      role="option"
    >
       
      <item>
        
         
        <p>
          Collapsed Node
        </p>
        
      </item>
    </node>
  </div>
</body>
`;

exports[`Node.svelte > should render node with expanded icon 1`] = `
<body>
  <div>
    <node
      aria-selected="false"
      class="min-h-5 flex flex-row cursor-pointer select-none items-center"
      data-path=""
      id="id-"
      role="option"
    >
       
      <item>
        
         
        <p>
          Expanded Node
        </p>
        
      </item>
    </node>
  </div>
</body>
`;

exports[`Node.svelte > should render node with path 1`] = `
<body>
  <div>
    <node
      aria-selected="false"
      class="min-h-5 flex flex-row cursor-pointer select-none items-center"
      data-path="0"
      id="id-0"
      role="option"
    >
       
      <item>
        
         
        <p>
          Example
        </p>
        
      </item>
    </node>
  </div>
</body>
`;
